lib/widgets/content.dart

```diff
@@ -1722,26 +1730,163 @@ class AvatarImage extends StatelessWidget {
+          end: 0,
+          bottom: 0,
+          child: PresenceCircle(
+            userId: userIdForPresence!,
+            size: presenceCircleSize,
+            backgroundColor: backgroundColor)),
+      ]);
+    }
+
     return SizedBox.square(
       dimension: size,
       child: ClipRRect(
         borderRadius: BorderRadius.all(Radius.circular(borderRadius)),
         clipBehavior: Clip.antiAlias,
-        child: child));
+        child: result));
+  }
+}
+
+/// The green or orange-gradient circle representing [PresenceStatus].
+///
+/// [backgroundColor] is required and must not be [Colors.transparent].
+/// It exists to match the background on which the avatar image is painted.
+///
+/// By default, nothing paints for a user in the "offline" status
+/// (i.e. a user without a [PresenceStatus]).
+/// Pass true for [explicitOffline] to paint a gray circle.
+class PresenceCircle extends StatefulWidget {
+  const PresenceCircle({
+    super.key,
+    required this.userId,
+    required this.size,
+    this.backgroundColor,
+    this.explicitOffline = false,
+  });
+
+  final int userId;
+  final double size;
+  final Color? backgroundColor;
+  final bool explicitOffline;
+
+  /// Creates a [WidgetSpan] with a [PresenceCircle], for use in rich text
+  /// before a user's name.
+  static InlineSpan asWidgetSpan({
+    required int userId,
+    required double fontSize,
+    required TextScaler textScaler,
+    Color? backgroundColor,
+  }) {
+    final size = textScaler.scale(fontSize) / 2;
+    return WidgetSpan(
+      alignment: PlaceholderAlignment.middle,
+      child: Padding(
+        padding: const EdgeInsetsDirectional.only(end: 4),
+        child: PresenceCircle(
+          userId: userId,
+          size: size,
+          backgroundColor: backgroundColor,
+          explicitOffline: true)));
+  }
+
+  @override
+  State<PresenceCircle> createState() => _PresenceCircleState();
+}
+
+class _PresenceCircleState extends State<PresenceCircle> with PerAccountStoreAwareStateMixin {
+  Presence? model;
+
+  @override
+  void onNewStore() {
+    model?.removeListener(_modelChanged);
+    model = PerAccountStoreWidget.of(context).presence
+      ..addListener(_modelChanged);
+  }
+
+  @override
+  void dispose() {
+    model!.removeListener(_modelChanged);
+    super.dispose();
+  }
+
+  void _modelChanged() {
+    setState(() {
+      // The actual state lives in [model].
+      // This method was called because that just changed.
+    });
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    final status = model!.presenceStatusForUser(
+      widget.userId, utcNow: ZulipBinding.instance.utcNow());
+    final designVariables = DesignVariables.of(context);
+    final effectiveBackgroundColor = widget.backgroundColor ?? designVariables.mainBackground;
+    assert(effectiveBackgroundColor != Colors.transparent);
+
+    Color? color;
+    LinearGradient? gradient;
+
+    switch (status) {
```

Comment:
nit: join these to the switch that sets them
```suggestion
    Color? color;
    LinearGradient? gradient;
    switch (status) {
```